src/main/java/org/radarcns/redcap/integration/MpIntegrator.java

```diff
@@ -0,0 +1,106 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.net.URL;
+import java.util.Map;
+
+/** Handler for updating Integrator Redcap form parameters. The input parameters are
+ *      described by {@link IntegrationData}.
+ * @see MpClient
+ */
+public class MpIntegrator {
+
+    private static final Logger Logger = LoggerFactory.getLogger(MpIntegrator.class);
+
+    private static final String SEPARATOR = "-";
+
+    private static final String HUMAN_READABLE_IDENTIFIER_KEY = "Human-readable-identifier";
+
+    private MpClient mpClient;
+
+    public MpIntegrator(MpClient mpClient) {
+        this.mpClient = mpClient;
+    }
+
+    /**
+     * Performs update of the subject on Management portal from the information
+     * sent from Redcap trigger. If subject does not exist, it creates a new one,
+     * otherwise it just returns the subject.
+     * TODO update the subject in case it exists
+     * @see MpClient
+     * @param redcapUrl the Redcap URL
+     * @param projectId the Redcap Project ID
+     * @param recordId the Redcap Record ID
+     * @return {@link Subject} from the Management Portal
+     */
+    public Subject performSubjectUpdateOnMp(URL redcapUrl, Integer projectId,
+                                             Integer recordId, Map<String, String> attributes) {
+        try {
+            Project project = mpClient.getProject(redcapUrl, projectId);
+            String radarWorkPackage = project.getWorkPackage().toUpperCase();
+            String location = project.getLocation().toUpperCase();
+            String humanReadableId = createHumanReadableId(radarWorkPackage, project.getId().toString(), location, recordId.toString());
+
+            Subject subject = subjectExistsUpdateElseCreate(redcapUrl, projectId, recordId, project, humanReadableId, attributes);
+            return subject;
+
+        } catch (NullPointerException exc) {
+            Logger.error("Project or Project attributes (Work Package, etc) cannot be null", exc);
+            throw new IllegalStateException("Project or Project attributes (Work Package, etc) in MP cannot be null.", exc);
+        } catch (Exception exc) {
+            Logger.error(exc.getMessage(), exc);
+            throw new IllegalStateException("Subject creation cannot be completed.", exc);
+        }
+    }
+
+    private Subject subjectExistsUpdateElseCreate(URL redcapUrl, Integer projectId, Integer recordId, Project project, String humanReadableId, Map<String, String> attributes){
+        try {
+            Subject subject = mpClient.getSubject(redcapUrl, projectId, recordId);
+            if(subject != null) {
+                Logger.info("Subject for Record Id: {} at {} is already available, updating...", recordId,
+                        redcapUrl);
+                attributes.put(HUMAN_READABLE_IDENTIFIER_KEY, humanReadableId);
+                subject.setAttributes(attributes);
+                subject = mpClient.updateSubject(subject);
```

Comment:
You can compare the attributes in the subject here, If they are same, no need to make an additional request to update the subject.